systemdspawner/systemd.py

```diff
@@ -76,73 +76,73 @@ async def start_transient_service(
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectoryPreserve=
+    #
+    properties.setdefault("RuntimeDirectory", unit_name)
     properties.setdefault("RuntimeDirectoryMode", "700")
-    # preserve runtime directories across restarts
-    # allows `systemctl restart` to load the env
     properties.setdefault("RuntimeDirectoryPreserve", "restart")
 
-    if properties:
-        for key, value in properties.items():
-            if isinstance(value, list):
-                run_cmd += [f"--property={key}={v}" for v in value]
-            else:
-                # A string!
-                run_cmd.append(f"--property={key}={value}")
+    # Ensure that out of memory killing of a process run inside the user server
+    # (systemd unit), such as a Jupyter kernel, doesn't result in stopping or
+    # killing the user server.
+    #
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.service.html#OOMPolicy=
+    #
+    properties.setdefault("OOMPolicy", "continue")
 
+    # Pass configured properties via systemd-run's --property flag
+    for key, value in properties.items():
+        if isinstance(value, list):
+            # The properties dictionary is allowed to have a list of values for
+            # each of its keys as a way of allowing the same key to be passed
+            # multiple times.
+            run_cmd += [f"--property={key}={v}" for v in value]
+        else:
+            # A string!
+            run_cmd.append(f"--property={key}={value}")
+
+    # Create and reference an environment variable file in the first
+    # RuntimeDirectory entry, which is a whitespace-separated list of directory
+    # names.
+    #
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=
+    #
     if environment_variables:
+        runtime_dir = os.path.join(RUN_ROOT, properties["RuntimeDirectory"].split()[0])
         environment_file = make_environment_file(
             runtime_dir, unit_name, environment_variables
         )
         run_cmd.append(f"--property=EnvironmentFile={environment_file}")
 
-    # Explicitly check if uid / gid are not None, since 0 is valid value for both
-    if uid is not None:
-        run_cmd += ["--uid", str(uid)]
-
-    if gid is not None:
-        run_cmd += ["--gid", str(gid)]
-
-    if slice is not None:
-        run_cmd += [f"--slice={slice}"]
-
-    # We unfortunately have to resort to doing cd with bash, since WorkingDirectory property
-    # of systemd units can't be set for transient units via systemd-run until systemd v227.
-    # Centos 7 has systemd 219, and will probably never upgrade - so we need to support them.
-    run_cmd += [
-        "/bin/bash",
-        "-c",
-        "cd {wd} && exec {cmd} {args}".format(
-            wd=shlex.quote(working_dir),
-            cmd=" ".join([shlex.quote(c) for c in cmd]),
-            args=" ".join([shlex.quote(a) for a in args]),
-        ),
```

Comment:
We won't need `shlex.quote(...)` any more right?

Reply:
That should be correct, yes